### PR TITLE
Add bpf_load_program, bpf_load_program_xattr APIs

### LINKED ISSUE #186: Ebpf API should provide option to load raw eBPF byte code
Ebpf API should provide option to load raw eBPF byte code from a file

### PR DESCRIPTION
**Pull request by dthaler**
This is a prerequisite for one of the other issues assigned to me for this month.

Fixes #186

Signed-off-by: Dave Thaler <dthaler@microsoft.com>

### CODE REVIEW

File: libs/api_common/api_common.hpp [hunk 1]
```diff
@@ -264,6 +264,10 @@ ebpf_result_to_errno(ebpf_result_t result)
         error = EEXIST;
         break;
 
+    case EBPF_VERIFICATION_FAILED:
+        error = EACCES;
```
Comment (shankarseal) [inline libs/api_common/api_common.hpp:268]: should this be documented as possible return value for the new APIs in libbpf.h ?

**Review by shankarseal — APPROVED** [review]

File: libs/api_common/api_common.hpp (hunk 1 above)
Comment (dthaler) [reply libs/api_common/api_common.hpp:268]: Yes, updated

File: libs/api/ebpf_api.cpp [hunk 2]
```diff
@@ -1717,10 +1717,68 @@ _ebpf_object_create_maps(_Inout_ ebpf_object_t* object)
...
+    load_info.object_name = const_cast<char*>(unique_name);
+    load_info.section_name = const_cast<char*>(unique_name);
+    load_info.program_name = const_cast<char*>(unique_name);
+    load_info.program_type = *program_type;
+    load_info.program_handle = reinterpret_cast<file_handle_t>(program_handle);
+    load_info.execution_type = execution_type;
+    load_info.byte_code = const_cast<uint8_t*>(byte_code);
+    load_info.byte_code_size = byte_code_size;
+    load_info.execution_context = execution_context_kernel_mode;
+    load_info.map_count = 0;
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:1756]: It is possible the program is accessing some maps. We will need to parse the instructions to find the map fds the program is using by looking at PSEUDO MAP FD flag in the instructions, and then get the map handles for those fds and pass fd to handle map in the RPC call.

Comment (dthaler) [reply libs/api/ebpf_api.cpp:1756]: If I add a TODO comment, can we track that as a separate issue?

File: libs/api/ebpf_api.cpp [hunk 3]
```diff
@@ -1717,10 +1717,68 @@ _ebpf_object_create_maps(_Inout_ ebpf_object_t* object)
...
+        if (log_buffer_output) {
+            strcpy_s(log_buffer, log_buffer_size, log_buffer_output);
+        }
+    }
+    ebpf_free_string(log_buffer_output);
+
+    if (result == EBPF_SUCCESS) {
+        *program_fd = _create_file_descriptor_for_handle(program_handle);
+    }
+    CloseHandle(program_handle);
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:1773]: I think we should close `program_handle` only in error case. Else the `program_fd` will become invalid.

File: libs/api/ebpf_api.cpp (hunk 2 above)
Comment (saxena-anurag) [reply libs/api/ebpf_api.cpp:1756]: Yes we can track it as a separate issue. 
`_resolve_maps_in_byte_code()` (in ebpfsvc) already does the same to get the map fds from the instructions.

File: libs/api/ebpf_api.cpp (hunk 3 above)
Comment (dthaler) [reply libs/api/ebpf_api.cpp:1773]: Right, done

File: libs/api/ebpf_api.cpp (hunk 2 above)
Comment (dthaler) [reply libs/api/ebpf_api.cpp:1756]: Filed issue and added TODO comment.

File: libs/api/ebpf_api.cpp (hunk 3 above)
Comment (saxena-anurag) [reply libs/api/ebpf_api.cpp:1773]: I see the new commit is pushed to "Add bpf() API" PR, but not this PR. Is that expected?

Comment (dthaler) [reply libs/api/ebpf_api.cpp]: No, i just fixed it :)

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**